### ISSUE TITLE
"-H key" to print config for just one top level key

### ISSUE THREAD

**Issue by andrus** [bootique/bootique#347]
Let's allow `-H` to take an optional argument that will be treated as a config path, so the command would limit printed config to just the children of that path

**Comment by carrollsa** [comment]
This seems at first glance to be a pretty good first issue. If that's the case, I'd like to take it on. And I figured I'd sent in the CLA when I make a PR. Can do it earlier if that's preferable.

**Comment by andrus** [comment]
Hi Steve, indeed. This is a rather isolated task that would not require and overall design changes. So go for it.

**Comment by carrollsa** [comment]
It's taken some time to get up to speed, and I've had some stuff going on. Wanted to update that I'm still working on this.

**Comment by andrus** [comment]
>  haven't done much OS, so please forgive me if these are things I should figure out for myself, but I'd love some input before I dive too far in.

Don't worry about this. Asking questions and communicting your ideas before sending PRs is a great thing and much appreciated.

> I don't want to jump into modifying CommandMetadata without good reason

Back to `-H`, good catch. We do not currently allow to pass arguments to commands. I think here we have a good example of why we should allow it. So I'd be +1 on adding the new functionality to `CommandMetadata`. Probably can borrow the the naming from `OptionMetadata.Builder`.

**Comment by carrollsa** [comment]
Thought I reacted to your comment. Sorry for going dark! Had a little time to work on this and focused on getting the `--help` printout to work, which involves converting the `CommandMetadata` to `OptionMetadata`. It was a little tricky to convert the value and cardinality in a way that didn't feel overwrought, and I may have gone too far that way. Tried a couple of routes that all seemed clunky or just weren't working for me, like explicitly getting and setting value and cardinality (got difficult when we didn't have a value). Curious what you think of the following (will also push a branch if that's easier to take a look).

```
// CommandMetadata
    public OptionMetadata asOption() {
        ...
        CommandValueCardinality cardinality = getValueCardinality();
        String valueName = getValueName();

        Consumer<OptionMetadata.Builder> optionsConsumer = builder -> {
            switch (cardinality) {
                case REQUIRED:
                    builder.valueRequired(valueName);
                    break;
                case OPTIONAL:
                    builder.valueOptional(valueName);
                    break;
                case NONE:
                    break;
                default:
                    throw new IllegalStateException("Unknown command value cardinality: " + cardinality);
            }
        };

        return OptionMetadata.builder(getName())
                .shortName(getShortName())
                .description(getDescription())
                .setValueWithCardinality(optionsConsumer)
                .build();
    }
```

```
// OptionMetadata
        Builder{
           ... 
           public Builder setValueWithCardinality(Consumer<Builder> optionsConsumer) {
                optionsConsumer.accept(this);
                return this;
           }
        }
```

**Comment by carrollsa** [comment]
There's something that's been puzzling me. In the weeds a bit but will try to sum things up.

By my understanding, we determine whether to call `Command.run()` with the following:
```
// JoptCliFactory
commandName(OptionSet optionSet) {
        Set<String> matches = new HashSet<>(3);
        getCommandManager().getAllCommands().forEach((name, mc) -> {
            if (!mc.isHidden() && !mc.isDefault() && optionSet.has(name) && !optionSet.hasArgument(name)) {
                matches.add(name);
            }
        });
    ...
}
...
```

I would think we would not want to negate `optionSet.hasArgument(name)` because by my reading, the current code is saying "There is no argument, so we do match and therefore execute `Command.run()`. As it stands, this works when calling `-H some_arg` (we do call `HelpConfigCommand.run()`), although I would think that if an arg is passed, `optionSet.hasArgument(name)` should be `true`.

Now that I've added `.valueOptional()` to the builder of `HelpConfigCommand`, `optionSet.hasArgument(name) == true` (which is what I would expect because there is an arg), and therefore we are not matching and not calling `Command.run()`.

I feel I must be missing something and have spent a decent amount of time trying to figure it out. If this doesn't ring a bell, I'll keep at it, but I wanted to see if it sticks out to you that I'm misinterpreting something. Thanks!

**Comment by andrus** [comment]
> I would think we would not want to negate optionSet.hasArgument(name) because by my reading, the current code is saying "There is no argument, so we do match and therefore execute Command.run().

I suspect the only reason for `!optionSet.hasArgument(name)` check is that currently we don't allow options with arguments for commands. Per this task we want to change that, so this check should go away.

**Comment by andrus** [comment]
> It was a little tricky to convert the value and cardinality in a way that didn't feel overwrought, and I may have gone too far that way.

I think you are on the right path. A few suggestions:

* You can reuse `OptionValueCardinality` instead of introducing `CommandValueCardinality`. I don't mind if command metatada would contain option-like attributes.
* `OptionMetadata.setValueWithCardinality` is non needed (also it breaks OptionMetadata immutability). You can apply the cardinality call to the mutable `OptionMetadata.Builder` instance. E.g., rewrite `createOptionsConsumer` as `OptionMetadata.Builder applyCardinality(OptionMetadata.Builder)`

**Comment by carrollsa** [comment]
I'm not sure we break the immutability of `OptionMetadata` as `setValueWithCardinality()` was on the `OptionMetadata.Builder`. Maybe using `set` in a Builder method confused things. Regardless, we can refactor to put `applyCardinality()` on the `OptionMetadata` class if that's preferable. My original intent was to not have any confusion from including both `valueRequired()/Optional()` methods and an `applyCardinality()` method on the same `Builder` and keep the `createConsumer()` as a private method on `CommandMetadata` because its use is only really in converting the `Command` to an `Option`.

If we do want to keep it all in `OptionMetadata.Builder`, is this what you had in mind? Your reply implies we'd want to pass an instance of `OptionMetadata.Builder` as the param to `applyCardinality`, but I couldn't see how that would work on a Builder, as won't have any existing values to pull from during the building process. I instead passed both `cardinality` and `valueName`. This also requires a check to make sure we aren't setting a value with a `NONE` cardinality since it's exposed now. Maybe I'm missing something!

```
        public Builder applyCardinality(OptionValueCardinality valueCardinality, String valueName) {
            switch (valueCardinality) {
                case REQUIRED:
                    return this.valueRequired(valueName);
                case OPTIONAL:
                    return this.valueOptional(valueName);
                case NONE:
                    if (valueName != null) {
                        throw new IllegalArgumentException("Value name cannot be provided for an option with cardinality NONE");
                    }

                    return this;
                default:
                    throw new IllegalStateException("Unknown command value cardinality: " + valueCardinality);
            }
        }
```

**Comment by carrollsa** [comment]
I moved forward with the above option for the time being. Now to the meat of the issue, are we opting to replace the current `-H` functionality from https://github.com/bootique/bootique/issues/267 which prints direct matches to the arg passed in? Or do we want it to be able to take both config paths and the previous direct matches of module names?

And some extra info, we did previously allow Commands to take args, but there wasn't anything set up for Command cardinality, and the existing functionality of `-h` couldn't account for optional args. The changes I've made to this point now allow Commands the same functionality as Options in that regard.

**Comment by andrus** [comment]
> I'm not sure we break the immutability of OptionMetadata as setValueWithCardinality() was on the OptionMetadata.Builder. 

Sorry, my bad. You are right. So that's just a question of naming, not mutability. 

> I wanted to clarify... we are opting to replace the current -H functionality from https://github.com/bootique/bootique/issues/267 which prints direct matches to the arg passed in, correct?

To be completely honest, I overlooked the presence #267 (as I knew that we don't allow arguments for commands, so I could not imagine how that would work). Now that I see it, indeed, I'd rather we reimplement it as a command with an optional cardinality. So the current task still needed, and we can also (separately) improve how it handles filtering.

**Comment by carrollsa** [comment]
I think I have a functioning implementation of Commands that take optional args. I pushed a branch up and have a draft PR open with some comments.

https://github.com/bootique/bootique/pull/348

A couple questions:

1. Can you provide an example input/output of what you mean by "children of [the input] path"? I have barely scratched the surface actual feature we're trying to add since I went to the Command arg implementation, but I'd like to make sure I fully understand before diving in.
2. Is there a place I could add a call to `ModuleMetadata.builder()....build()` to register a simple module so I can do some debugging or would it take some more? I figured it would be simplest to do this in a test to step through things, but I was having trouble.

**Comment by andrus** [comment]
> Do we want to handle the update to CommandMetadata separately from this issue?

Yes, you can open a PR with a description for just the CommandMetadata piece, so that we have a tracking number that we can put in RELEASE-NOTES

> I don't see any tests specifically for OptionMetadata but do see some tests in Bootique_CliOptionsIT that check interpretation of CLI args with regard to Options. I'm guessing I should make some similar tests for the new Command functionality, but do we want any new testing for Command beyond that?

Let's start with tests in `Bootique_CliOptionsIT`

> I noticed that our predicate currently filters on matching the config name, when I believe https://github.com/bootique/bootique/issues/267 intended for it to filter by module name.

No, this was done correctly. The filter should apply to configs. Still we can imporve it in two ways:

* exclude modules from the output whose configuration got filtered out
* work on hierarcical filtering (config paths like "a.b.c" should narrow down the configs by walkign the tree).

However, let's do it one step at a time. I suggest we focus on the Command-related part now. And we will review the filtering after that.

**Comment by carrollsa** [comment]
Have been out of town and pretty swamped when home. Trying to get to this this week.

**Comment by andrus** [comment]
Closing this task. It is working more or less, as parts of it were handled by #267 and #348 PR. We can still improve filtering per may earlier comments:

>  exclude modules from the output whose configuration got filtered out
>  work on hierarchical filtering (config paths like "a.b.c" should narrow down the configs by walking the tree).

But we should probably open a separate more specific task for those improvements
